plugin.go

```diff
@@ -0,0 +1,190 @@
+package sendremotefile
+
+import (
+	"context"
+	"errors"
+	"io"
+	"net/http"
+	"strings"
+	"sync"
+	"time"
+
+	"go.uber.org/zap"
+)
+
+const (
+	PluginName        string = "sendremotefile"
+	ContentTypeKey    string = "Content-Type"
+	ContentTypeVal    string = "application/octet-stream"
+	xSendRemoteHeader string = "X-Sendremotefile"
+	bufSize           int    = 10 * 1024 * 1024 // 10MB chunks
+)
+
+type Logger interface {
+	NamedLogger(name string) *zap.Logger
+}
+
+type Plugin struct {
+	log         *zap.Logger
+	writersPool sync.Pool
+}
+
+func (p *Plugin) Init(log Logger) error {
+	p.log = log.NamedLogger(PluginName)
+
+	p.writersPool = sync.Pool{
+		New: func() any {
+			wr := new(writer)
+			wr.code = http.StatusOK
+			wr.data = make([]byte, 0, 10)
+			wr.hdrToSend = make(map[string][]string, 2)
+			return wr
+		},
+	}
+
+	return nil
+}
+
+func (p *Plugin) Middleware(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		rrWriter := p.getWriter()
+		defer func() {
+			p.putWriter(rrWriter)
+			_ = r.Body.Close()
+		}()
+
+		next.ServeHTTP(rrWriter, r)
+
+		// if there is no X-Sendremotefile header from the PHP worker, just return
+		if path := rrWriter.Header().Get(xSendRemoteHeader); path == "" { //nolint:nestif
+			for k := range rrWriter.hdrToSend {
+				for kk := range rrWriter.hdrToSend[k] {
+					// re-add all headers from the worker
+					w.Header().Add(k, rrWriter.hdrToSend[k][kk])
+				}
+			}
+
+			// write original
+			w.WriteHeader(rrWriter.code)
+			if len(rrWriter.data) > 0 {
+				// write a body if exists
+				_, err := w.Write(rrWriter.data)
+				if err != nil {
+					p.log.Error("failed to write data to the response", zap.Error(err))
+				}
+			}
+
+			return
+		}
+
+		// we already checked that that header exists
+		path := rrWriter.Header().Get(xSendRemoteHeader)
+		// delete the original X-Sendremotefile header
+		rrWriter.Header().Del(xSendRemoteHeader)
+
+		// re-add original headers
+		for k := range rrWriter.hdrToSend {
+			for kk := range rrWriter.hdrToSend[k] {
+				// re-add all headers from the worker
+				w.Header().Add(k, rrWriter.hdrToSend[k][kk])
+			}
+		}
+
+		if !strings.HasPrefix(path, "http") {
+			http.NotFound(w, r)
+			return
+		}
+
+		ctx, cancel := context.WithCancel(context.TODO())
+		timer := time.AfterFunc(5*time.Second, func() {
+			cancel()
+		})
+
+		req, err := http.NewRequest("GET", path, nil)
```

Comment:
Would be better to separate the logic of getting the file to the separate method.